rskj-core/src/main/java/org/ethereum/vm/PrecompiledContracts.java

```diff
@@ -422,4 +453,209 @@ private BigInteger parseArg(byte[] data, int offset, int len) {
 
     }
 
+    private static byte[] encodeRes(byte[] w1, byte[] w2) {
+
+        byte[] res = new byte[64];
+
+        w1 = stripLeadingZeroes(w1);
+        w2 = stripLeadingZeroes(w2);
+
+        System.arraycopy(w1, 0, res, 32 - w1.length, w1.length);
+        System.arraycopy(w2, 0, res, 64 - w2.length, w2.length);
+
+        return res;
+    }
+
+    /**
+     * Computes point addition on Barreto–Naehrig curve.
+     * See {@link BN128Fp} for details<br/>
+     * <br/>
+     *
+     * input data[]:<br/>
+     * two points encoded as (x, y), where x and y are 32-byte left-padded integers,<br/>
+     * if input is shorter than expected, it's assumed to be right-padded with zero bytes<br/>
+     * <br/>
+     *
+     * output:<br/>
+     * resulting point (x', y'), where x and y encoded as 32-byte left-padded integers<br/>
+     *
+     */
+    public static class BN128Addition extends PrecompiledContract {
```

Comment:
I think it would be better if you extract the `BN128Addition`, `BN128Multiplication` and `BN128Pairing` contracts to their own files

Reply:
Do you want me to changed the implementation to the new NativeContract interface? That way they can be on its own class also

Reply:
I'm happy with just moving it outside the `org.ethereum.vm.PrecompiledContracts`, if you want to give a try to the `NativeContract` interface go ahead :)